File: javascript/src/agent/reviewer/reviewerUnits.test.ts
// Deterministic unit tests for the reviewer's pure helpers: findings parsing/normalization, the
// exclude filter, the size gate, category selection, the count-based scorecard, the glue gates, and
// the coalesce/enqueue math. No model calls, no network — canned inputs only.
import { describe, expect, it } from 'vitest';

import type { PRFile } from '../../githubapi/client';
import { newEnvelope, Kind } from '../../ingest/envelope';
import { CATEGORIES, hasUiFiles, selectCategories } from './categories';
import { coalesceKey, enqueueOptions } from './enqueue';
import {
  Dimension,
  Severity,
  clampConfidence,
  clampThreshold,
  fingerprint,
  findingsJson,
  newFinding,
  normalizeDimension,
  normalizeMessage,
  normalizeSeverity,
  parseFindings,
  severityRank,
} from './findings';
import { AVG_DIFF_LINE_BYTES, FileFilter, globToRegExp, patchBytes } from './filter';
import { dedupe, demoteToNitpick, dropLowConfidence } from './glue';
import { Level, dimLevel, levelGlyph, levelWord, scoreFindings } from './scorecard';
import { oversize } from './sizegate';

function file(partial: Partial<PRFile>): PRFile {
  return { path: '', previousPath: '', status: '', additions: 0, deletions: 0, patch: '', ...partial };
}

describe('findings parsing', () => {
  it('extracts a JSON array wrapped in fences and prose', () => {
    const raw = 'Here you go:\n```json\n[{"file":"a.go","line":3,"dimension":"security","severity":"critical","message":"sqli","confidence":0.9}]\n```\nDone.';
    const out = parseFindings(raw);
    expect(out).toHaveLength(1);
    expect(out[0]!.file).toBe('a.go');
    expect(out[0]!.dimension).toBe(Dimension.Security);
    expect(out[0]!.severity).toBe(Severity.Critical);
    expect(out[0]!.confidence).toBe(0.9);
  });

  it('treats a malformed body as no findings', () => {
    expect(parseFindings('not json at all')).toEqual([]);
    expect(parseFindings('[unterminated')).toEqual([]);
    expect(parseFindings('')).toEqual([]);
  });

  it('skips an empty array in favor of a following populated one', () => {
    const raw = 'first [] then [{"file":"b.go","message":"note"}]';
    const out = parseFindings(raw);
    expect(out).toHaveLength(1);
    expect(out[0]!.file).toBe('b.go');
  });

  it('drops a finding with no message', () => {
    expect(parseFindings('[{"file":"a.go","line":1}]')).toEqual([]);
  });

  it('rejects a non-finite confidence in the validation layer', () => {
    // NaN/Infinity are not valid JSON tokens; a string "NaN" fails the number check, so the whole
    // array is rejected and the scan finds no other array — no findings.
    expect(parseFindings('[{"message":"x","confidence":"NaN"}]')).toEqual([]);
    expect(parseFindings('[{"message":"x","confidence":true}]')).toEqual([]);
  });

  it('rejects a non-integer line in the validation layer', () => {
    expect(parseFindings('[{"message":"x","line":1.5}]')).toEqual([]);
  });

  it('normalizes unknown severity and dimension to safe defaults', () => {
    const out = parseFindings('[{"message":"x","severity":"bogus","dimension":"weird"}]');
    expect(out[0]!.severity).toBe(Severity.Nitpick);
    expect(out[0]!.dimension).toBe(Dimension.Other);
  });
});

describe('findings normalization + fingerprint', () => {
  it('folds spaces and hyphens in a dimension', () => {
    expect(normalizeDimension('Runtime Safety')).toBe(Dimension.RuntimeSafety);
    expect(normalizeDimension('error-handling')).toBe(Dimension.ErrorHandling);
  });

  it('maps known severities and defaults the rest to nitpick', () => {
    expect(normalizeSeverity('CRITICAL')).toBe(Severity.Critical);
    expect(normalizeSeverity('')).toBe(Severity.Nitpick);
  });

  it('orders severities so worse ranks higher', () => {
    expect(severityRank(Severity.Critical)).toBeGreaterThan(severityRank(Severity.Major));
    expect(severityRank(Severity.Medium)).toBeGreaterThan(severityRank(Severity.Nitpick));
  });

  it('fingerprints on file+line+normalized message, ignoring dimension', () => {
    const a = newFinding({ file: 'a.go', line: 5, message: 'Nil  Deref', dimension: Dimension.Security });
    const b = newFinding({ file: 'a.go', line: 5, message: 'nil deref', dimension: Dimension.RuntimeSafety });
    expect(fingerprint(a)).toBe(fingerprint(b));
    expect(normalizeMessage('  Foo   BAR ')).toBe('foo bar');
  });

  it('renders findings as compact JSON with the wire keys', () => {
    const json = findingsJson([newFinding({ file: 'a.go', fixPrompt: 'fix', ruleId: 'R1' })]);
    expect(json).toContain('"fix_prompt":"fix"');
    expect(json).toContain('"rule_id":"R1"');
    expect(json).not.toContain(' '); // compact
  });
});

describe('confidence clamps', () => {
  it('clampConfidence treats zero/absent as 0.5 and has NO NaN guard', () => {
    expect(clampConfidence(0)).toBe(0.5);
    expect(clampConfidence(-1)).toBe(0.5);
    expect(clampConfidence(2)).toBe(1);
    expect(clampConfidence(0.7)).toBe(0.7);
    expect(Number.isNaN(clampConfidence(NaN))).toBe(true); // no NaN guard
  });

  it('clampThreshold folds NaN and negatives to 0 (keep all)', () => {
    expect(clampThreshold(NaN)).toBe(0);
    expect(clampThreshold(-1)).toBe(0);
    expect(clampThreshold(2)).toBe(1);
    expect(clampThreshold(0.6)).toBe(0.6);
  });
});

describe('exclude filter', () => {
  it('compiles globs and matches basenames vs full paths', () => {
    const f = new FileFilter(['*.min.js', 'vendor/**', 'go.sum', '']);
    expect(f.excluded('a/b/app.min.js')).toBe(true);
    expect(f.excluded('vendor/x/y.go')).toBe(true);
    expect(f.excluded('go.sum')).toBe(true);
    expect(f.excluded('src/main.go')).toBe(false);
  });

  it('totals patch bytes on the kept set only', () => {
    const { kept, diffBytes } = new FileFilter(['go.sum']).apply([
      file({ path: 'main.go', patch: '12345' }),
      file({ path: 'go.sum', patch: 'ignored' }),
    ]);
    expect(kept).toHaveLength(1);
    expect(diffBytes).toBe(5);
  });

  it('charges an omitted patch from its line counts, binaries nothing', () => {
    expect(patchBytes(file({ path: 'a', patch: '', additions: 2, deletions: 1 }))).toBe(3 * AVG_DIFF_LINE_BYTES);
    expect(patchBytes(file({ path: 'logo.png', patch: '' }))).toBe(0);
  });

  it('escapes regexp metacharacters and treats ** vs * distinctly', () => {
    expect(globToRegExp('a.b').test('axb')).toBe(false); // '.' is literal
    expect(globToRegExp('a.b').test('a.b')).toBe(true);
    expect(globToRegExp('*.go').test('x/y.go')).toBe(false); // '*' does not cross '/'
    expect(globToRegExp('**.go').test('x/y.go')).toBe(true); // '**' crosses '/'
  });
});

describe('size gate', () => {
  it('denies over either dimension and disables a non-positive cap', () => {
    expect(oversize(60, 10, 50, 0).denied).toBe(true);
    expect(oversize(1, 2000, 50, 1000).denied).toBe(true);
    expect(oversize(1, 10, 50, 1000).denied).toBe(false);
    expect(oversize(9999, 9999, 0, 0).denied).toBe(false); // both dims disabled
  });
});

describe('category selection', () => {
  it('gates accessibility to UI/markup diffs', () => {
    expect(hasUiFiles([file({ path: 'a.go' })])).toBe(false);
    expect(hasUiFiles([file({ path: 'ui/App.tsx' })])).toBe(true);
    const noUi = selectCategories([file({ path: 'a.go' })]).map((c) => c.name);
    expect(noUi).not.toContain('accessibility');
    const withUi = selectCategories([file({ path: 'a.go' }), file({ path: 'x.css' })]).map((c) => c.name);
    expect(withUi).toContain('accessibility');
  });

  it('always includes the other catch-all and never marks a dotfile as UI', () => {
    expect(CATEGORIES.some((c) => c.other)).toBe(true);
    expect(hasUiFiles([file({ path: '.tsx' })])).toBe(false); // leading dot = no extension
  });
});

describe('scorecard', () => {
  it('derives per-dimension levels from severity counts', () => {
    expect(dimLevel({ dimension: Dimension.Security, critical: 1, major: 0, medium: 0, nitpick: 0, level: Level.Green })).toBe(Level.Red);
    expect(dimLevel({ dimension: Dimension.Security, critical: 0, major: 2, medium: 0, nitpick: 0, level: Level.Green })).toBe(Level.Red);
    expect(dimLevel({ dimension: Dimension.Security, critical: 0, major: 1, medium: 0, nitpick: 0, level: Level.Green })).toBe(Level.Yellow);
    expect(dimLevel({ dimension: Dimension.Security, critical: 0, major: 0, medium: 3, nitpick: 0, level: Level.Green })).toBe(Level.Yellow);
    expect(dimLevel({ dimension: Dimension.Security, critical: 0, major: 0, medium: 1, nitpick: 9, level: Level.Green })).toBe(Level.Green);
  });

  it('caps overall to red on any critical in a critical dimension', () => {
    const card = scoreFindings([
      newFinding({ dimension: Dimension.Security, severity: Severity.Critical, message: 'x' }),
    ]);
    expect(card.overall).toBe(Level.Red);
    expect(card.total).toBe(1);
  });

  it('otherwise takes the worst dimension level (no critical-dim cap)', () => {
    // A single major in a non-critical dimension is yellow, not red.
    const card = scoreFindings([
      newFinding({ dimension: Dimension.Performance, severity: Severity.Major, message: 'x' }),
    ]);
    expect(card.overall).toBe(Level.Yellow);
    expect(card.dims).toHaveLength(1);
  });

  it('renders level glyphs and words', () => {
    expect(levelGlyph(Level.Red)).toBe('🔴');
    expect(levelGlyph(Level.Yellow)).toBe('🟡');
    expect(levelGlyph(Level.Green)).toBe('🟢');
    expect(levelWord(Level.Red)).toBe('Red');
    expect(levelWord(Level.Yellow)).toBe('Yellow');
    expect(levelWord(Level.Green)).toBe('Green');
  });
});

describe('glue gates', () => {
  it('drops below the minimum confidence, keeps all when non-positive', () => {
    const fs = [newFinding({ message: 'a', confidence: 0.4 }), newFinding({ message: 'b', confidence: 0.8 })];
    expect(dropLowConfidence(fs, 0.6)).toHaveLength(1);
    expect(dropLowConfidence(fs, 0)).toHaveLength(2);
  });

  it('collapses duplicates by fingerprint keeping the worst severity', () => {
    const merged = dedupe([
      newFinding({ file: 'a.go', line: 1, message: 'x', severity: Severity.Nitpick, confidence: 0.9 }),
      newFinding({ file: 'a.go', line: 1, message: 'x', severity: Severity.Critical, confidence: 0.5 }),
    ]);
    expect(merged).toHaveLength(1);
    expect(merged[0]!.severity).toBe(Severity.Critical);
  });

  it('breaks a severity tie by higher confidence', () => {
    const merged = dedupe([
      newFinding({ file: 'a.go', line: 1, message: 'x', severity: Severity.Major, confidence: 0.5 }),
      newFinding({ file: 'a.go', line: 1, message: 'x', severity: Severity.Major, confidence: 0.9 }),
    ]);
    expect(merged[0]!.confidence).toBe(0.9);
  });

  it('demotes every finding to nitpick', () => {
    const out = demoteToNitpick([newFinding({ message: 'x', severity: Severity.Critical })]);
    expect(out[0]!.severity).toBe(Severity.Nitpick);
  });
});

describe('enqueue coalescing', () => {
  function reviewEnvelope(action: string, at: Date) {
    const body = `{"action":"${action}","pull_request":{"number":7,"head":{"ref":"x","sha":"s"}},"repository":{"full_name":"acme/web.api"}}`;
    return newEnvelope(Kind.Review, 'webhook:/github', Buffer.from(body), at);
  }

  it('debounces a synchronize push under a per-PR dedup name', () => {
    const opts = enqueueOptions(reviewEnvelope('synchronize', new Date(1_700_000_000_000)), 30_000);
    expect(opts.delayMs).toBe(30_000);
    expect(opts.name).toMatch(/^review-[A-Za-z0-9_-]+-7-\d+$/);
  });

  it('coalesces pushes in one window and separates a later window', () => {
    const base = 1_700_000_000_000;
    const a = enqueueOptions(reviewEnvelope('synchronize', new Date(base + 2_000)), 30_000);
    const b = enqueueOptions(reviewEnvelope('synchronize', new Date(base + 5_000)), 30_000);
    const c = enqueueOptions(reviewEnvelope('synchronize', new Date(base + 45_000)), 30_000);
    expect(a.name).toBe(b.name);
    expect(a.name).not.toBe(c.name);
  });

  it('enqueues opened/reopened/ready_for_review immediately', () => {
    for (const action of ['opened', 'reopened', 'ready_for_review']) {
      expect(enqueueOptions(reviewEnvelope(action, new Date(0)), 30_000)).toEqual({});
    }
  });

  it('yields no options for a non-review kind, unparseable payload, or non-positive debounce', () => {
    const notReview = newEnvelope(Kind.CI, 'webhook:/github', Buffer.from('{}'), new Date(0));
    expect(enqueueOptions(notReview, 30_000)).toEqual({});
    const bad = newEnvelope(Kind.Review, 'webhook:/github', Buffer.from('{not json'), new Date(0));
    expect(enqueueOptions(bad, 30_000)).toEqual({});
    expect(enqueueOptions(reviewEnvelope('synchronize', new Date(0)), 0)).toEqual({});
  });

  it('encodes the repo losslessly so near-identical repos do not collide', () => {
    const bucket = 1_700_000_000_000_000_000n;
    const a = coalesceKey({ action: '', number: 7, repoFullName: 'acme/web.api', headRef: '', headSha: '', baseRef: '', draft: false, labels: [], authorLogin: '' }, bucket);
    const b = coalesceKey({ action: '', number: 7, repoFullName: 'acme/web-api', headRef: '', headSha: '', baseRef: '', draft: false, labels: [], authorLogin: '' }, bucket);
    expect(a).not.toBe(b);
  });
});

File: kotlin/src/test/kotlin/com/automation/agent/agent/reviewer/ReviewerUnitsTest.kt
package com.automation.agent.agent.reviewer

import com.automation.agent.githubapi.PRFile
import com.automation.agent.ingest.Envelope
import com.automation.agent.ingest.Kind
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import java.time.Instant
import kotlin.time.Duration.Companion.seconds

private fun file(path: String, patch: String = "", status: String = "modified", add: Int = 0, del: Int = 0) =
    PRFile(path = path, status = status, additions = add, deletions = del, patch = patch)

class ReviewerUnitsTest : BehaviorSpec({
    Given("the defensive findings parser") {
        When("the model wraps a JSON array in prose and fences") {
            Then("it recovers the findings") {
                val raw = "Here you go:\n```json\n[{\"file\":\"a.kt\",\"line\":3,\"dimension\":\"security\"," +
                    "\"severity\":\"critical\",\"message\":\"secret\",\"confidence\":0.9}]\n```\ndone"
                val found = parseFindings(raw)
                found.size shouldBe 1
                found[0].file shouldBe "a.kt"
                found[0].line shouldBe 3
                found[0].dimension shouldBe Dimension.SECURITY
                found[0].severity shouldBe Severity.CRITICAL
                found[0].confidence shouldBe 0.9
            }
        }
        When("an earlier non-findings array precedes the real one") {
            Then("it skips to the first decodable findings array") {
                val found = parseFindings("[1,2,3] then [{\"message\":\"hi\"}]")
                found.size shouldBe 1
                found[0].message shouldBe "hi"
            }
        }
        When("the body is malformed or empty") {
            Then("it degrades to no findings") {
                parseFindings("not json at all") shouldBe emptyList()
                parseFindings("[]") shouldBe emptyList()
                parseFindings("[{\"message\":\"\"}]") shouldBe emptyList() // empty message is unusable
            }
        }
        When("a confidence is non-finite") {
            Then("the array is rejected at validation (no findings)") {
                parseFindings("[{\"message\":\"x\",\"confidence\":1e999}]") shouldBe emptyList()
            }
        }
        When("a known string field has the wrong type") {
            Then("the array is rejected") {
                parseFindings("[{\"message\":\"x\",\"file\":123}]") shouldBe emptyList()
            }
        }
    }

    Given("severity and dimension normalization") {
        Then("unknown or blank values fall to the safe buckets") {
            normalizeSeverity("CRITICAL") shouldBe Severity.CRITICAL
            normalizeSeverity("bogus") shouldBe Severity.NITPICK
            normalizeSeverity("") shouldBe Severity.NITPICK
            normalizeDimension("Runtime-Safety") shouldBe Dimension.RUNTIME_SAFETY
            normalizeDimension("weird space") shouldBe Dimension.OTHER
            severityRank(Severity.CRITICAL) shouldBe 4
        }
    }

    Given("the fingerprint") {
        Then("it folds whitespace/case and omits the dimension") {
            val a = Finding(file = "a.kt", line = 5, dimension = Dimension.SECURITY, message = "Same   Message")
            val b = Finding(file = "a.kt", line = 5, dimension = Dimension.PERFORMANCE, message = "same message")
            fingerprint(a) shouldBe fingerprint(b)
        }
    }

    Given("the confidence clamps") {
        Then("the threshold folds NaN/negative to 0 and caps at 1") {
            clampThreshold(Double.NaN) shouldBe 0.0
            clampThreshold(-1.0) shouldBe 0.0
            clampThreshold(2.0) shouldBe 1.0
            clampThreshold(0.6) shouldBe 0.6
        }
        Then("the value clamp treats 0/absent as 0.5 and caps at 1") {
            clampConfidence(0.0) shouldBe 0.5
            clampConfidence(2.0) shouldBe 1.0
            clampConfidence(0.7) shouldBe 0.7
        }
    }

    Given("the two-dimensional size gate") {
        Then("it denies over the file cap or the byte cap and passes otherwise") {
            oversize(60, 10, 50, 262144).denied shouldBe true
            oversize(5, 300000, 50, 262144).denied shouldBe true
            oversize(5, 100, 50, 262144).denied shouldBe false
            oversize(9999, 9999999, 0, 0).denied shouldBe false // non-positive caps disable a dimension
        }
    }

    Given("the exclude-glob filter") {
        val filter = FileFilter(listOf("*.min.js", "vendor/**", "go.sum", ""))
        Then("basename globs, path globs, and ** all match") {
            filter.excluded("app.min.js") shouldBe true
            filter.excluded("a/b/app.min.js") shouldBe true
            filter.excluded("vendor/x/y.go") shouldBe true
            filter.excluded("go.sum") shouldBe true
            filter.excluded("src/main.kt") shouldBe false
        }
        Then("apply keeps non-excluded files and totals the patch bytes on the filtered set") {
            val result = filter.apply(listOf(file("app.min.js", "x".repeat(9999)), file("a.kt", "abc")))
            result.kept.map { it.path } shouldBe listOf("a.kt")
            result.diffBytes shouldBe 3
        }
    }

    Given("patch-byte accounting") {
        Then("it uses the patch length, estimates an omitted text diff, and charges binary as zero") {
            patchBytes(file("a.kt", "hello")) shouldBe 5
            patchBytes(file("big.kt", "", add = 3, del = 1)) shouldBe 4 * AVG_DIFF_LINE_BYTES
            patchBytes(file("img.png", "")) shouldBe 0
        }
    }

    Given("category selection") {
        Then("accessibility is gated to a UI/markup change") {
            selectCategories(listOf(file("a.kt"))).map { it.name } shouldBe
                listOf("safety", "security", "performance", "code_quality", "other")
            selectCategories(listOf(file("a.kt"), file("page.tsx"))).any { it.name == "accessibility" } shouldBe true
            hasUiFiles(listOf(file("styles.css"))) shouldBe true
        }
    }

    Given("the count-based scorecard") {
        Then("a critical in a critical dimension caps the overall to red") {
            val card = scoreFindings(
                listOf(
                    Finding(dimension = Dimension.SECURITY, severity = Severity.CRITICAL, message = "s"),
                    Finding(dimension = Dimension.READABILITY, severity = Severity.MEDIUM, message = "r"),
                ),
            )
            card.overall shouldBe Level.RED
            card.total shouldBe 2
        }
        Then("dimension thresholds and glyphs are as documented") {
            dimLevel(DimScore(Dimension.PERFORMANCE, major = 2)) shouldBe Level.RED
            dimLevel(DimScore(Dimension.PERFORMANCE, major = 1)) shouldBe Level.YELLOW
            dimLevel(DimScore(Dimension.PERFORMANCE, medium = 3)) shouldBe Level.YELLOW
            dimLevel(DimScore(Dimension.PERFORMANCE)) shouldBe Level.GREEN
            levelWord(Level.RED) shouldBe "Red"
            levelGlyph(Level.GREEN).isNotEmpty() shouldBe true
        }
    }

    Given("the deterministic glue gates") {
        Then("dropLowConfidence, dedupe, and demoteToNitpick behave") {
            val low = Finding(file = "a", line = 1, message = "m", confidence = 0.2)
            val high = Finding(file = "a", line = 1, message = "m", severity = Severity.MAJOR, confidence = 0.9)
            dropLowConfidence(listOf(low, high), 0.6).map { it.confidence } shouldBe listOf(0.9)
            // Same fingerprint, worst severity kept.
            val crit = Finding(file = "a", line = 1, message = "m", severity = Severity.CRITICAL, confidence = 0.5)
            dedupe(listOf(high, crit)).let {
                it.size shouldBe 1
                it[0].severity shouldBe Severity.CRITICAL
            }
            demoteToNitpick(listOf(crit)).all { it.severity == Severity.NITPICK } shouldBe true
        }
    }

    Given("diff formatting") {
        Then("it fences each file's patch and notes an omitted diff") {
            val diff = formatDiff(listOf(file("a.kt", "line", status = "modified"), file("img.png", "", status = "added")))
            diff shouldContain "### a.kt (modified)"
            diff shouldContain "```diff"
            diff shouldContain "(no textual diff available)"
        }
        Then("a patch with backtick runs gets a longer fence") {
            maxBacktickRun("a ``` b ```` c") shouldBe 4
            formatDiff(listOf(file("md.md", "text ``` more"))) shouldContain "````diff"
        }
    }

    Given("the intake helpers") {
        Then("splitFullName and isDependencyBot behave") {
            splitFullName("acme/api").let { it.ok shouldBe true; it.owner shouldBe "acme"; it.repo shouldBe "api" }
            splitFullName("nope").ok shouldBe false
            splitFullName("a/b/c").ok shouldBe false
            isDependencyBot("dependabot[bot]") shouldBe true
            isDependencyBot("renovate[bot]") shouldBe true
            isDependencyBot("alice") shouldBe false
        }
    }

    Given("instruction composition") {
        Then("the review and glue instructions embed the diff (and prior findings)") {
            buildReviewInstruction("PROMPT", "DIFF", null) shouldContain "## Diff under review"
            val glue = buildGlueInstruction("PROMPT", "DIFF", listOf(Finding(file = "a", line = 1, message = "m")), null)
            glue shouldContain "## Findings already reported by other lenses"
            glue shouldContain "\"file\":\"a\""
        }
    }

    Given("the debounce/coalesce hints") {
        val payload = "{\"action\":\"synchronize\",\"number\":42,\"repository\":{\"full_name\":\"acme/api\"}," +
            "\"pull_request\":{\"number\":42}}"
        Then("a synchronize review yields a byte-identical per-PR-per-window dedup name") {
            val e = Envelope.new(Kind.REVIEW, "webhook:/github", payload.toByteArray(), Instant.ofEpochSecond(1_000_000_000))
            val opts = enqueueOptions(e, 30.seconds)
            opts.name shouldBe "review-YWNtZS9hcGk-42-999999990000000000"
            opts.delay shouldBe 30.seconds
        }
        Then("a non-review kind, a non-synchronize action, or a non-positive debounce yields no hints") {
            val lint = Envelope.new(Kind.LINT, "webhook:/lint", payload.toByteArray(), Instant.EPOCH)
            enqueueOptions(lint, 30.seconds).name shouldBe null
            val opened = "{\"action\":\"opened\",\"number\":1,\"repository\":{\"full_name\":\"a/b\"},\"pull_request\":{\"number\":1}}"
            val e = Envelope.new(Kind.REVIEW, "webhook:/github", opened.toByteArray(), Instant.EPOCH)
            enqueueOptions(e, 30.seconds).name shouldBe null
            val sync = Envelope.new(Kind.REVIEW, "webhook:/github", payload.toByteArray(), Instant.EPOCH)
            enqueueOptions(sync, kotlin.time.Duration.ZERO).name shouldBe null
        }
    }

    Given("findingsJson") {
        Then("it renders the wire keys") {
            val js = findingsJson(listOf(Finding(file = "a.kt", line = 2, dimension = Dimension.SECURITY, message = "m", confidence = 0.5)))
            js shouldContain "\"dimension\":\"security\""
            js shouldContain "\"fix_prompt\":\"\""
        }
    }
})
